lightning/src/ln/channel.rs

```diff
@@ -1360,6 +1375,243 @@ pub(super) struct ChannelContext<SP: Deref> where SP::Target: SignerProvider {
 	/// If we can't release a [`ChannelMonitorUpdate`] until some external action completes, we
 	/// store it here and only release it to the `ChannelManager` once it asks for it.
 	blocked_monitor_updates: Vec<PendingChannelMonitorUpdate>,
+	// If we've sent `commtiment_signed` for an interactive transaction construction,
+	// but have not received `tx_signatures` we MUST set `next_funding_txid` to the
+	// txid of that interactive transaction, else we MUST NOT set it.
+	next_funding_txid: Option<Txid>,
+}
+
+#[cfg(any(dual_funding, splicing))]
+pub(super) trait InteractivelyFunded<SP: Deref> where SP::Target: SignerProvider {
+	fn context(&self) -> &ChannelContext<SP>;
+
+	fn context_mut(&mut self) -> &mut ChannelContext<SP>;
+
+	fn interactive_tx_constructor_mut(&mut self) -> &mut Option<InteractiveTxConstructor>;
+
+	fn dual_funding_context(&self) -> &DualFundingChannelContext;
+
+	fn dual_funding_context_mut(&mut self) -> &mut DualFundingChannelContext;
+
+	fn is_initiator(&self) -> bool;
+
+	fn begin_interactive_funding_tx_construction<ES: Deref>(
+		&mut self, signer_provider: &SP, entropy_source: &ES, holder_node_id: PublicKey,
+	) -> Result<Option<InteractiveTxMessageSend>, APIError>
+	where ES::Target: EntropySource
+	{
+		let mut funding_inputs_prev_outputs: Vec<TxOut> = Vec::with_capacity(self.dual_funding_context_mut().our_funding_inputs.len());
+		// Check that vouts exist for each TxIn in provided transactions.
+		for (idx, input) in self.dual_funding_context_mut().our_funding_inputs.iter().enumerate() {
+			if let Some(output) = input.1.as_transaction().output.get(input.0.previous_output.vout as usize) {
+				funding_inputs_prev_outputs.push(output.clone());
+			} else {
+				return Err(APIError::APIMisuseError {
+					err: format!("Transaction with txid {} does not have an output with vout of {} corresponding to TxIn at funding_inputs[{}]",
+						input.1.as_transaction().txid(), input.0.previous_output.vout, idx) });
+			}
+		}
+		let total_input_satoshis: u64 = self.dual_funding_context_mut().our_funding_inputs.iter().map(
+			|input| input.1.as_transaction().output[input.0.previous_output.vout as usize].value.to_sat()).sum();
+		if total_input_satoshis < self.dual_funding_context_mut().our_funding_satoshis {
```

Comment:
The `total_input_satoshis` just counts the total of all the prevout amounts that the user will add to the funding transaction, while `our_funding_satohis` is the contribution to the funding output. The difference is used for fees + change.

Reply:
It's possible I left something in here from an old revision of the PR. Thanks, I'll have a look at what's going on.

Reply:
So I checked and, yes, fees are calculated in `maybe_add_funding_change_output` but we still don't check if we have enough input value to cover what we want to add to the funding output (i.e. `our_funding_satoshis`) other than on this line. I still feel we need it here rather than putting it in `maybe_add_funding_change_output` to fail early and tell the user they didn't provide enough input value to even cover what they want to contribute.

Reply:
It's a good point but I reckoned we'd provide the flexibility if the user of the API had some larger value UTXOs they wanted to split up. If we're happy that they wouldn't have that flexibility and would need to have the right value UTXOs from the start, it would definitely simplify a lot for us as we'd never bother generating change and just use everything for their contribution.

I don't have a strong opinion. We can always add flexibility later so I'm happy to just use the entire input value minus fees.